### ISSUE TITLE
Investigate workflow job failing: Typecheck/ESLint actions out of memory

### ISSUE THREAD

**Issue by blazejkustra** [Expensify/App#48470]
[Typecheck action that failed](https://github.com/Expensify/App/actions/runs/10682355263/job/29607993149?pr=47468)

Some context from couple months ago when ESLint was failing: https://github.com/Expensify/App/issues/44425

<details><summary>Issue Owner</summary>Current Issue Owner: @blazejkustra</details>

**Comment by blazejkustra** [comment]
I asked @roryabraham for assignment on #typescript-new-expensify channel, waiting for his approval for now.

**Comment by roryabraham** [comment]
🟢 from me (as discussed in slack) to try and figure out how to make lint and tsc faster

**Comment by blazejkustra** [comment]
# Coming back with some findings 🔍 

TS server is quite fast in general, so any slowness is usually caused by the devs or complex libraries, common culprits are:
- huge union types, 
- deep recursive type, 
- complex generic types,
- massive files.

Currently, TS check is at around 54 seconds for me:

<details>
<summary>Baseline ~54s, ~4.5GB of memory</summary>

```
NODE_OPTIONS=--max_old_space_size=8192 tsc --extendedDiagnostics --incremental false
```

```
Files:                         5765
Lines of Library:             40640
Lines of Definitions:        510909
Lines of TypeScript:         353141
Lines of JavaScript:           7532
Lines of JSON:                 1267
Lines of Other:                   0
Identifiers:                 976989
Symbols:                    3123303
Types:                      2321176
Instantiations:            17286341
Memory used:               4503960K
Assignability cache size:   2050430
Identity cache size:          40895
Subtype cache size:           41483
Strict subtype cache size:   384382
I/O Read time:                0.83s
Parse time:                   1.15s
ResolveModule time:           0.53s
ResolveTypeReference time:    0.01s
ResolveLibrary time:          0.00s
Program time:                 2.86s
Bind time:                    0.61s
Check time:                  49.83s
I/O Write time:               0.00s
printTime time:               0.08s
Emit time:                    0.08s
Total time:                  53.38s
```

</details>

## Investigation

That got me thinking what types are so complicated that E/App takes so long to compile... I looked at the AST tree and saw a lot of onyx symbols used:
```
  // run a script to count all references
  [ '__type', 213203 ],
  [ 'BaseMappingStringKeyAndSelector', 190291 ], <- onyx
  [ 'BaseMappingFunctionKeyAndSelector', 190291 ], <- onyx
  [ 'Selector', 185896 ], <- onyx
  [ 'Mapping', 178545 ], <- onyx
  [ '__object', 52927 ],
  [ 'RelativeIndexable', 39501 ],
  [ '__jsxAttributes', 39028 ],
  [ 'Array', 35424 ],
  [ 'CollectionMapping', 25118 ], <- onyx
  [ 'G', 20500 ],  <- possibly onyx
  [ 'at', 16904 ],
  [ '__function', 15337 ],
  [ 'String', 10594 ],
  [ 'U', 8215 ],
  [ 'S', 8134 ],
```
These symbols are used in `withOnyx` HOC, so I went ahead and prepared a script to automatically migrate most of them to use `useOnyx` hook instead, and run typecheck with diagnostics again:

<details>
<summary>~38s, ~2.8GB of memory (40% less memory used 🚀 )</summary>

```
Files:                         5769
Lines of Library:             41147
Lines of Definitions:        510910
Lines of TypeScript:         355488
Lines of JavaScript:           7532
Lines of JSON:                 1269
Lines of Other:                   0
Identifiers:                 979940
Symbols:                    2573629
Types:                      1208298
Instantiations:             5587137
Memory used:               2856605K
Assignability cache size:   1335744
Identity cache size:          22177
Subtype cache size:           41927
Strict subtype cache size:   373653
I/O Read time:                1.20s
Parse time:                   1.74s
ResolveModule time:           0.62s
ResolveTypeReference time:    0.01s
ResolveLibrary time:          0.01s
Program time:                 3.97s
Bind time:                    0.84s
Check time:                  33.85s
printTime time:               0.00s
Emit time:                    0.00s
Total time:                  38.67s
```

</details>

[Here](https://github.com/Expensify/App/pull/48647) is the PR where most of `withOnyx` are migrated, but there are still around 40 HOCs in use so 38s is not the final score! 🤟 

## Other things to consider
- Split big files into smaller, especially these three:
├─ Check file /users/blazejkustra/documents/swmansion/expensify/swmapp/src/libs/actions/iou.ts (2134ms)
├─ Check file /users/blazejkustra/documents/swmansion/expensify/swmapp/src/libs/actions/policy/policy.ts (1161ms)
├─ Check file /users/blazejkustra/documents/swmansion/expensify/swmapp/src/libs/actions/report.ts (997ms)
  - intellisense, prettier, lint, ts are super slow in these files, splitting will help to decrease calculations while editing 

- Calculating translations is a heavy operation which creates a big union which slows down TS and ESLint
![image](https://github.com/user-attachments/assets/6c6c45b6-b603-475b-bda8-b0c6d9b4e5a5)
  - we should consider rewriting it a bit, maybe instead 'common.address' keys we should use an array key ['common', 'address'] or other approach - needs more investigation
- Caching `tsbuildinfo` in Github actions (we already do it locally, `--incremental` flag)
- Split the project into smaller pieces with [typescript references](https://www.typescriptlang.org/docs/handbook/project-references.html) 
  - this needs more investigations, I don't expect massive gains with this approach as it's hard to split E/App into independent modules

Other general solutions, but not worth the effort for now:
- Use interfaces instead of types (tested it locally, didn't see much difference)
- Split complex types into smaller (this helps typescript to cache them better)

<details>
<summary>Hot spots, what files takes most time to compile - we should investigate them 👀</summary>

Command:
```
NODE_OPTIONS=--max_old_space_size=8192 tsc --generateTrace trace --incremental false && open trace
npx @typescript/analyze-trace trace
```

Hot Spots:
```
├─ Check file /users/blazejkustra/documents/swmansion/expensify/swmapp/src/pages/workspace/accounting/netsuite/import/netsuiteimportmappingpage.tsx (651ms)
├─ Check file /users/blazejkustra/documents/swmansion/expensify/swmapp/src/libs/errorutils.ts (620ms)
└─ Check file /users/blazejkustra/documents/swmansion/expensify/swmapp/src/pages/settings/profile/personaldetails/stateselectionpage.tsx (614ms)
</details>
```

## Materials
For anybody interested, here're some useful links:
- https://github.com/microsoft/TypeScript-wiki/blob/main/Performance-Tracing.md
- https://www.youtube.com/watch?v=IzzvGrZrFSE
- https://www.edgedb.com/blog/an-approach-to-optimizing-typescript-type-checking-performance#how-do-we-quantify-the-type-checker-s-workload
- https://www.youtube.com/watch?v=ZL3z1oBZntk&t=348s

**Comment by blazejkustra** [comment]
# TLDR;

1. **To improve the performance of TS and decrease memory usage we should start migrating `withOnyx` to `useOnyx`.**  
3. Rethink our translation system as types are currently too complex
4. Split big files into smaller modules (`actions/IOU.ts`, `actions/Report.ts`, `actions/Policy.ts`)
5. Investigate these three files, for some reason TS is quite slow there:
```
├─ Check file /src/pages/workspace/accounting/netsuite/import/netsuiteimportmappingpage.tsx (651ms)
├─ Check file /src/libs/errorutils.ts (620ms)
└─ Check file /src/pages/settings/profile/personaldetails/stateselectionpage.tsx (614ms)
```

**Comment by blazejkustra** [comment]
I'll be monitoring the performance of TS/ESLint once we got to a point where there are far less `withOnyx` in the codebase. Let's leave the issue open for now.

**Comment by blazejkustra** [comment]
Once there are no withOnyx left, I'll post an update to compare TS/ESlint speed
